### PR TITLE
Sort friends a to z

### LINKED ISSUE #246: Move friends into it's own route and mock alpha sorted list simmilar to uplink mobile.

### LINKED ISSUE #280: Friend List should be displayed in alphabetical order
**Issue:** Like title says, when you go to friends page the list of friends should be displayed in Alphabetical Order 

**Screen recording:**

https://user-images.githubusercontent.com/93608357/200979917-019ecd09-c0d3-4b5d-be51-c55de97687e4.mov

### PR DESCRIPTION
**Pull request by lgmarchi**
### 1. Sort friend list by first char from username 

https://user-images.githubusercontent.com/63157656/202223373-43a1c397-3a8b-4523-9697-4a81333faf43.mov

### 2. Friend component receive already the username now, instead of search for it 
image

### 3. Receive friend did list as before, but get all username, re order alphabetically and return it to render
a. The struct now used on render is FriendListAlpha
image

b. All the time to update list alpha, we use a struct to save friend did and username, called UsernameAndDID 
image

c. Main part of the function that order friends list
image

d. The friend list is rebuilt if occurs some change on each 300 miliseconds
image

e. To render friends list separated per char now, we use to maps on rsx 
image

### 4. Alpha list to navigate 
image

image

a. Used ul with li and a elements 
b. Prevent default parameter avoid a common problem : "Some javascript code is sending an AJAX request (post or get) that is preventing the link from functioning as expected". 
c. Used rel parameter on anchor tags to avoid security problems 
image

### Which issue(s) this PR fixes 🔨
- Resolve #246

- Resolve #280

### CODE REVIEW

**Comment by phillsatellite** [comment]
Hi Lucas! Found a small weird issue on this pr, whenever I try to add a user with the letter Z in the beginning of the name they wont appear after I accept request. I tried replicating this on dev branch but could not 

Dev branch 

https://user-images.githubusercontent.com/93608357/202235676-18a17503-e6a4-45c8-928f-37561675e0c0.mov

This branch 

https://user-images.githubusercontent.com/93608357/202235701-dfc59ab8-de9a-42a5-8cf2-072c9a83ea3f.mov

**Comment by lgmarchi** [comment]
> Hi Lucas! Found a small weird issue on this pr, whenever I try to add a user with the letter Z in the beginning of the name they wont appear after I accept request. I tried replicating this on dev branch but could not
> 
> Dev branch
> 
>  dev.mov 
> This branch
> 
>  branch.mov

Good Catch Phill, I am checking right now

**Review by phillsatellite — APPROVED** [review]

**Comment by InfamousVague** [comment]
This looks good but I wonder if we can build the list items in a easier to read way, something like 
```rs
let alpha: Vec<_> = "abcdefghijklmnopqrstuvwxyz".chars().collect();
for char in &alpha {
  rsx! { li { a { href: "#{char.to_string()}", prevent_default: "onclick", rel: "noopener noreferrer", char.to_string() } }
}
```

we may want to also consider instead building the alphabetic list programmatically so we can support multiple languages (maybe building it based off the first character in each of your friends names) but this is a future ticket.

**Comment by lgmarchi** [comment]
> This looks good but I wonder if we can build the list items in a easier to read way, something like
> 
> ```rust
> let alpha: Vec<_> = "abcdefghijklmnopqrstuvwxyz".chars().collect();
> for char in &alpha {
>   rsx! { li { a { href: "#{char.to_string()}", prevent_default: "onclick", rel: "noopener noreferrer", char.to_string() } }
> }
> ```
> 
> we may want to also consider instead building the alphabetic list programmatically so we can support multiple languages (maybe building it based off the first character in each of your friends names) but this is a future ticket.

Awesome, I was thinking to do that and improve the code, thanks Matt, changes pushed.

File: src/components/main/friends/mod.rs [hunk 1]
```diff
@@ -21,20 +33,25 @@ pub struct Props {
 pub fn Friends(cx: Scope<Props>) -> Element {
     log::debug!("rendering Friends");
     let add_error = use_state(&cx, String::new);
-    let friends = use_state(&cx, || {
-        HashSet::from_iter(cx.props.account.list_friends().unwrap_or_default())
-    });
+    let friends_grouped_per_first_letter = use_state(&cx, || Vec::new());
```
Comment (dariusc93) [inline src/components/main/friends/mod.rs]:
```suggestion
    let friends_grouped_per_first_letter = use_state(&cx, Vec::new);
```

File: src/components/main/friends/mod.rs [hunk 2]
```diff
@@ -21,20 +33,25 @@ pub struct Props {
...
+    let friends = use_state(&cx, || HashSet::new());
```
Comment (dariusc93) [inline src/components/main/friends/mod.rs]:
```suggestion
    let friends = use_state(&cx, HashSet::new);
```

File: src/components/main/friends/mod.rs [hunk 3]
```diff
@@ -21,20 +33,25 @@ pub struct Props {
...
+            &cx.props.account.clone(),
+            friends_grouped_per_first_letter,
+        ),
+        |(friends, mp, friends_grouped_per_first_letter)| async move {
             loop {
                 let friends_list: HashSet<_> =
                     HashSet::from_iter(mp.read().list_friends().unwrap_or_default());
 
-                if *friends != friends_list {
+                if *friends != friends_list && friends_list.len() > 0 {
```
Comment (dariusc93) [inline src/components/main/friends/mod.rs]:
```suggestion
                if *friends != friends_list && !friends_list.is_empty() {
```

File: src/components/main/friends/mod.rs [hunk 4]
```diff
@@ -21,20 +33,25 @@ pub struct Props {
...
                     log::debug!("updating friends list ");
+                    let new_friends_list = order_friend_list(&friends_list, &mp);
```
Comment (dariusc93) [inline src/components/main/friends/mod.rs]:
```suggestion
                    let new_friends_list = order_friend_list(&friends_list, &mp).unwrap_or_default();
```

File: src/components/main/friends/mod.rs [hunk 5]
```diff
@@ -51,19 +73,114 @@ pub fn Friends(cx: Scope<Props>) -> Element {
...
                 }
             }
-        }
     })
 }
+
+fn order_friend_list(
+    friend_did_list: &HashSet<warp::crypto::DID>,
+    account: &Account,
+) -> Vec<FriendListAlpha> {
```
Comment (dariusc93) [inline src/components/main/friends/mod.rs]:
```suggestion
) -> Option<Vec<FriendListAlpha>> {
```

File: src/components/main/friends/mod.rs [hunk 6]
```diff
@@ -51,19 +73,114 @@ pub fn Friends(cx: Scope<Props>) -> Element {
...
+            username: _friend_username,
+            did: friend_did.clone(),
+        };
+        username_did.push(_friend_username_and_did);
+    }
+
+    username_did.sort_by(|a, b| a.username.cmp(&b.username));
+
+    // Get the first letter username that start the list
+    let mut old_letter: char = username_did[0]
```
Comment (dariusc93) [inline src/components/main/friends/mod.rs]:
```suggestion
    let mut old_letter: char = username_did.first()?
```

File: src/components/main/friends/mod.rs [hunk 7]
```diff
@@ -51,19 +73,114 @@ pub fn Friends(cx: Scope<Props>) -> Element {
...
+        .username
+        .to_uppercase()
+        .chars()
+        .next()
```
Comment (dariusc93) [inline src/components/main/friends/mod.rs]:
```suggestion
        .next()?;
```

File: src/components/main/friends/mod.rs [hunk 8]
```diff
@@ -51,19 +73,114 @@ pub fn Friends(cx: Scope<Props>) -> Element {
...
+        .unwrap();
```
Comment (dariusc93) [inline src/components/main/friends/mod.rs]:
```suggestion
```

File: src/components/main/friends/mod.rs [hunk 9]
```diff
@@ -51,19 +73,114 @@ pub fn Friends(cx: Scope<Props>) -> Element {
...
+
+    // Group friends per first username letter
+    for (_friend, is_last_friend) in username_did
+        .iter()
+        .enumerate()
+        .map(|(i, f)| (f, i == username_did.len() - 1))
+    {
+        let first_letter_friend_username = _friend.username.to_uppercase().chars().next().unwrap();
```
Comment (dariusc93) [inline src/components/main/friends/mod.rs]:
```suggestion
        let first_letter_friend_username = _friend.username.to_uppercase().chars().next()?;
```

**Review by dariusc93 — COMMENTED** [review]
Looks good so far. Some things that I saw yesterday already been approve but I did make some suggestions just to reduce the direct unwrapping

**Comment by lgmarchi** [comment]
> Looks good so far. Some things that I saw yesterday already been approve but I did make some suggestions just to reduce the direct unwrapping

Thanks Bro, changes made and pushed.

**Comment by sdwoodbury** [comment]
@lgmarchi  I pushed some changes that I hope simplify the logic a little

**Review by sdwoodbury — APPROVED** [review]

**Comment by lgmarchi** [comment]
> @lgmarchi I pushed some changes that I hope simplify the logic a little

Thanks Stu, nice changes. Tested again and worked with no problem.

**Pull request merged**